dbt/models/reporting/reporting.ratio_stats.py

```diff
@@ -4,14 +4,462 @@
+
+    if len(out) > 1:
+        raise Exception("".join(map(str, out)))
+
+
+# Formula specific bootstrapping functions
+def cod_boot(ratio, nboot=100, alpha=0.05):
+    return boot_ci(cod, ratio=ratio, nboot=nboot, alpha=alpha)
+
+
+def prd_boot(assessed, sale_price, nboot=100, alpha=0.05):
+    return boot_ci(
+        prd, assessed=assessed, sale_price=sale_price, nboot=nboot, alpha=alpha
+    )
+
+
+def boot_ci(fun, nboot=100, alpha=0.05, **kwargs):
+    """
+    Calculate the non-parametric bootstrap confidence interval
+    for a given numeric input and a chosen function.
+
+    :param fun:
+        Function to bootstrap. Must return a single value.
+    :param nboot:
+        Default 100. Number of iterations to use to estimate
+        the output statistic confidence interval.
+    :param alpha:
+        Default 0.05. Numeric value indicating the confidence
+        interval to return. 0.05 will return the 95% confidence interval.
+    :param kwargs:
+        Arguments passed on to ``fun``.
+    :type fun: function
+    :type nboot: int
+    :type alpha: float
+    :type kwargs: numeric
+
+    .. note::
+       Input function should require 1 argument or be ``assesspy.prd()``.
+
+    :return:
+        A two-long list of floats containing the bootstrapped confidence
+        interval of the input vector(s).
+    :rtype: list[float]
+
+    :Example:
+
+    .. code-block:: python
+
+        # Calculate PRD confidence interval:
+        import assesspy as ap
+
+        ap.boot_ci(
+            ap.prd,
+            assessed = ap.ratios_sample().assessed,
+            sale_price = ap.ratios_sample().sale_price,
+            nboot = 100
+            )
+    """
+
+    # Make sure prd is passed arguments in correct order
+    if fun.__name__ == "prd" and set(["assessed", "sale_price"]).issubset(
+        kwargs.keys()
+    ):
+        kwargs = (kwargs["assessed"], kwargs["sale_price"])
+    elif fun.__name__ == "prd" and not set(
+        ["assessed", "sale_price"]
+    ).issubset(kwargs.keys()):
+        raise Exception(
+            "PRD function expects argurments 'assessed' and 'sale_price'."
+        )
+    else:
+        kwargs = tuple(kwargs.values())
+
+    check_inputs(kwargs)  # Input checking and error handling
+
+    num_kwargs = len(kwargs)
+    kwargs = pd.DataFrame(kwargs).T
+    n = len(kwargs)
+
+    # Check that the input function returns a numeric vector
+    out = (
+        fun(kwargs.iloc[:, 0])
+        if num_kwargs < 2
+        else fun(kwargs.iloc[:, 0], kwargs.iloc[:, 1])
+    )
+    if not is_numeric_dtype(out):
+        raise Exception("Input function outputs non-numeric datatype.")
+
+    data_array = kwargs.to_numpy()
+
+    def bootstrap_worker(
+        data_array, fun, num_kwargs, n, nboot, start, end, result_queue
+    ):
+        ests = []
+        for _ in range(start, end):
+            sample_indices = np.random.choice(
+                data_array.shape[0], size=n, replace=True
+            )
```

Comment:
We substitute the old pandas sampling for a faster `np.random.choice()` sampling.